nntrainer/include/databuffer_v2.h

```diff
@@ -0,0 +1,325 @@
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug		No known bugs except for NYI items
+ * @brief	This is databuffer class for Neural Network
+ *
+ * @todo TODO: Support multi files for dataset with files
+ * @todo TODO: Support multi threads with more than 1 thread and use thread
+ * pooling
+ * @todo TODO: Support label size to be 0 for inference based scenarios
+ * @todo TODO: rename data buffer to dataset
+ * @todo TODO: manage with just 1 buffer
+ * @todo TODO: consider different data structure for buffer when using shuffle
+ */
+
+#ifndef __DATABUFFER_V2_H__
+#define __DATABUFFER_V2_H__
+#ifdef __cplusplus
+
+#include <condition_variable>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <tuple>
+#include <vector>
+
+#include <nntrainer-api-common.h>
+#include <tensor.h>
+
+namespace nntrainer {
+
+/**
+ * @brief   States the collect thread running in the background
+ */
+enum class ThreadStates {
+  THREAD_NULL,            /**< not yet initialized */
+  THREAD_READY,           /**< initialized but not yet started */
+  THREAD_RUNNING,         /**< started and running */
+  THREAD_REQUEST_TO_STOP, /**< main thread has request to stop, background
+                             thread is in the process of being stopped */
+  THREAD_STOPPED, /**< background thread has stopped but not yet cleaned up */
+  THREAD_EPOCH_FINISHED, /**< background threads have finished the epoch */
+  THREAD_ERROR           /**< an error has occurred in background threads */
+};
+
+/**
+ * @brief   Dataset generator callback type declaration
+ */
+typedef std::function<std::remove_pointer<ml_train_datagen_cb>::type>
+  datagen_cb;
+
+/**
+ * @class   DataBuffer Data Buffers
+ * @brief   Data Buffer for read and manage data
+ */
+class DataBuffer_v2 {
+public:
+  /**
+   * @brief     Constructor
+   */
+  DataBuffer_v2() :
+    type(DataBufferType::DATA_BUFFER_UNKNOWN),
+    buffer_len(1),
+    avail_buffer_idx(0),
+    total_data_entries(0),
+    batch_size(1),
+    num_threads(1),
+    generator(nullptr),
+    gen_user_data(nullptr),
+    thread_state(ThreadStates::THREAD_NULL) {
+    buffer.clear();
+    batched_buffer.clear();
+    label_size.resize(1, 0);
+    input_size.resize(1, 0);
+  }
+
+  /**
+   * @brief     Destructor
+   */
+  ~DataBuffer_v2() { stop(); }
+
+  /**
+   * @brief     Initialize Buffer with set properties
+   * @throws std::invalid_argument
+   * @throws std::runtime_error
+   */
+  void init();
+
+  /**
+   * @brief     start the thread for collection the data
+   * @throws std::runtime_error
+   */
+  void start();
+
+  /**
+   * @brief     function for thread to stop collection of the data
+   * @throws std::runtime_error
+   */
+  void stop();
+
+  /**
+   * @brief     Get data from buffer to the passed input and label vectors
+   * @param[in] inputs list of input tensors
+   * @param[in] inputs list of label tensors
```

Comment:
It's `getData` and where is the location of data stored? Perhaps, `@param[out]`?

Reply:
Thanks for the review.
Currently this patch has been abandoned as it is low priority.
This will be taken back in a couple of weeks. Then I will update as per your suggestion.